### PR TITLE
Stabilize `const_slice_split_at_mut` and `const_slice_first_last_chunk`

### LINKED ISSUE #101804: Tracking Issue for `feature(const_slice_split_at_mut)`
Feature gate: `#![feature(const_slice_split_at_mut)]`

This is a tracking issue for spliting mutable slices in const context.

### Public API

```rust
// slice
pub const fn split_at_mut(&mut self, mid: usize) -> (&mut [T], &mut [T]);
pub const fn split_at_mut_unchecked(&mut self, mid: usize) -> (&mut [T], &mut [T]);
pub const fn split_at_mut_checked(&mut self, mid: usize) -> Option<(&mut [T], &mut [T])>;
```

### Steps / History

- [x] Implementation: #101800
- [x] https://github.com/rust-lang/rust/pull/124678
- [x] Wait for https://github.com/rust-lang/rust/issues/57349 to be stabilized.
- [ ] Wait for #130403 since from_raw_parts_mut is used internally
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR #130428

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

### LINKED ISSUE #111774: Tracking Issue for `slice_first_last_chunk` feature (`slice::{split_,}{first,last}_chunk{,_mut}`)
Feature gate: `#![feature(slice_first_last_chunk)]`

This is a tracking issue for the `slice::{split_,}{first,last}_chunk{,_mut}` API.

### Public API

2024-09-06: Most of this is stable now, except for the const-ness of the `mut` methods.

```rust
impl [T] {
    pub const fn first_chunk<const N: usize>(&self) -> Option<&[T; N]>;
    pub const fn first_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]>;
    pub const fn last_chunk<const N: usize>(&self) -> Option<&[T; N]>;
    pub const fn last_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]>;
    pub const fn split_first_chunk<const N: usize>(&self) -> Option<(&[T; N], &[T])>;
    pub const fn split_first_chunk_mut<const N: usize>(&mut self) -> Option<(&mut [T; N], &mut [T])>;
    pub const fn split_last_chunk<const N: usize>(&self) -> Option<(&[T; N], &[T])>;
    pub const fn split_last_chunk_mut<const N: usize>(&mut self) -> Option<(&mut [T; N], &mut [T])>;
}
```

Note that this mirrors the existing, already-stabilised API:

```rust
impl [T] {
    pub const fn first(&self) -> Option<&T>;
    pub const fn first_mut(&mut self) -> Option<&mut T>;
    pub const fn last(&self) -> Option<&T>;
    pub const fn last_mut(&mut self) -> Option<&mut T>;
    pub const fn split_first(&self) -> Option<(&T, &[T])>;
    pub const fn split_first_mut(&mut self) -> Option<(&mut T, &mut [T])>;
    pub const fn split_last(&self) -> Option<(&T, &[T])>;
    pub const fn split_last_mut(&mut self) -> Option<(&mut T, &mut [T])>;
}
```

### Steps / History

- [x] API change proposal (ACP)[^1]: rust-lang/libs-team#69
- [x] Implementation: #95198
- [ ] Final comment period (FCP)[^2]
- [ ] Wait until https://github.com/rust-lang/rust/issues/101804 is stable
- [ ] Stabilization PR

### Unresolved Questions

- [ ] Is this the best order for the tuple in `split_last_chunk`?  Consistency with `split_last` would have it be `-> Option<(&[T;N], &[T])>` as was ACPed, but it might be confusing to have the two slice-like things in the opposite order from every other API I can think of that returns two subslices of the original.  In https://github.com/rust-lang/rust/pull/95198#discussion_r1204824761, @scottmcm expected `-> Option<(&[T;N], &[T])>` for `split_first_chunk`, but for it to be `-> Option<(&[T], &[T;N])>` in `split_last_chunk` so the tuple is "in order", in some sense.
- ...

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html
[^2]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

### PR DESCRIPTION
**Pull request by onestacked**
Stabilizes #101804 and the remainder of #111774.

FCP proposed in the tracking issue.

Requires #130403 (or it would need a rustc_allow_const_fn_unstable for it)

Stabilized const API:
```rust
// slice
impl [T] {
    pub const fn split_at_mut(&mut self, mid: usize) -> (&mut [T], &mut [T]);
    pub const fn split_at_mut_unchecked(&mut self, mid: usize) -> (&mut [T], &mut [T]);
    pub const fn split_at_mut_checked(&mut self, mid: usize) -> Option<(&mut [T], &mut [T])>;

    pub const fn first_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]>;
    pub const fn last_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]>;
    pub const fn split_first_chunk_mut<const N: usize>(&mut self) -> Option<(&mut [T; N], &mut [T])>;
    pub const fn split_last_chunk_mut<const N: usize>(&mut self) -> Option<(&mut [T; N], &mut [T])>;
}
```

Closes #101804
Closes #111774

cc @RalfJung

### CODE REVIEW

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-const-eval

**Comment by workingjubilee** [comment]
r? @workingjubilee

@onestacked Could you remove the TODO and fix this up so it can be merged after the other PR lands? We usually just tag PRs as `S-blocked` to indicate they shouldn't be merged. You might have to fix it up for merge conflicts anyways.

File: library/core/src/slice/mod.rs [hunk 1]
```diff
@@ -2001,7 +2002,9 @@ impl<T> [T] {
     /// assert_eq!(v, [1, 2, 3, 4, 5, 6]);
     /// ```
     #[stable(feature = "slice_split_at_unchecked", since = "1.79.0")]
-    #[rustc_const_unstable(feature = "const_slice_split_at_mut", issue = "101804")]
+    #[rustc_const_stable(feature = "const_slice_split_at_mut", since = "CURRENT_RUSTC_VERSION")]
+    #[cfg_attr(bootstrap, rustc_allow_const_fn_unstable(const_mut_refs))]
+    #[rustc_allow_const_fn_unstable(const_slice_from_raw_parts_mut)] // TODO Wait for the Stabilization before merging!
```
Comment (workingjubilee) [inline library/core/src/slice/mod.rs:2007]: specifically
```suggestion
```

**Comment by onestacked** [comment]
@workingjubilee I'll wait until the FCP is done, to avoid having to rebase again and again.

**Comment by RalfJung** [comment]
If you want you can include stabilizing https://github.com/rust-lang/rust/issues/111774 in this PR as well, as it depends on the feature here and the two went into FCP together. Given the amount of feature gates moving towards stabilization, there's no reason to have an individual PR for each of them. :)

**Comment by workingjubilee** [comment]
Thank you!

File: library/core/src/slice/mod.rs [hunk 2]
```diff
@@ -384,6 +385,7 @@ impl<T> [T] {
     #[inline]
     #[stable(feature = "slice_first_last_chunk", since = "1.77.0")]
     #[rustc_const_stable(feature = "slice_first_last_chunk", since = "1.77.0")]
+    #[cfg_attr(bootstrap, rustc_allow_const_fn_unstable(const_mut_refs))]
```
Comment (RalfJung) [inline library/core/src/slice/mod.rs:388]: Why did you add this attribute? This function was already stable before...

Happens again below.

Comment (onestacked) [reply library/core/src/slice/mod.rs:388]: Since const_mut_refs is a compiler feature the bootstrap build needs to enable the feature (since the current beta compiler doesn't have it stabilized)

Error if removed:
```
   Compiling std v0.0.0 (/home/chrisi/Documents/code/rust/rust/library/std)
error: const-stable function cannot use `#[feature(const_mut_refs)]`
   --> core/src/slice/mod.rs:358:50
    |
358 |     pub const fn first_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]> {
    |                                                  ^^^^^^^^^
    |
help: if the function is not (yet) meant to be stable, make this function unstably const
    |
358 +     #[rustc_const_unstable(feature = "...", issue = "...")]
359 |     pub const fn first_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]> {
    |
help: otherwise, as a last resort `#[rustc_allow_const_fn_unstable]` can be used to bypass stability checks (but requires team approval)
    |
358 +     #[rustc_allow_const_fn_unstable(const_mut_refs)]
359 |     pub const fn first_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]> {
    |

error: const-stable function cannot use `#[feature(const_mut_refs)]`
   --> core/src/slice/mod.rs:365:27
    |
365 |             Some(unsafe { &mut *(self.as_mut_ptr().cast::<[T; N]>()) })
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
help: if the function is not (yet) meant to be stable, make this function unstably const
    |
358 +     #[rustc_const_unstable(feature = "...", issue = "...")]
359 |     pub const fn first_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]> {
    |
help: otherwise, as a last resort `#[rustc_allow_const_fn_unstable]` can be used to bypass stability checks (but requires team approval)
    |
358 +     #[rustc_allow_const_fn_unstable(const_mut_refs)]
359 |     pub const fn first_chunk_mut<const N: usize>(&mut self) -> Option<&mut [T; N]> {
    |
```

Comment (RalfJung) [reply library/core/src/slice/mod.rs:388]: Yes, `first_chunk_mut` needs this feature. But `first_chunk` does not.

Comment (onestacked) [reply library/core/src/slice/mod.rs:388]: Ah right sorry. Yeah I don't know why I added this. Removing it.

**Comment by RalfJung** [comment]
Both FCPs have passed, so this is ready to go.

@onestacked can you resolve the conflicts?

**Comment by RalfJung** [comment]
r=me when CI is green.

@bors delegate+

**Comment by onestacked** [comment]
CI passed.
@bors r=@RalfJung
